File: mini-diffusion-go/diffusion/unet.go
package diffusion

import (
	"math/rand"
)

// ResBlock is a residual block with time embedding.
type ResBlock struct {
	Conv1    *Conv2d
	Conv2    *Conv2d
	Norm1    *GroupNorm
	Norm2    *GroupNorm
	TimeMLP  *Linear
	SkipConv *Conv2d
	Channels int
}

// NewResBlock creates a new residual block.
func NewResBlock(inChannels, outChannels, timeEmbDim int, rng *rand.Rand) *ResBlock {
	var skipConv *Conv2d
	if inChannels != outChannels {
		skipConv = NewConv2d(inChannels, outChannels, 1, 1, 0, rng)
	}

	numGroups := 8
	if outChannels < 8 {
		numGroups = outChannels
	}

	return &ResBlock{
		Conv1:    NewConv2d(inChannels, outChannels, 3, 1, 1, rng),
		Conv2:    NewConv2d(outChannels, outChannels, 3, 1, 1, rng),
		Norm1:    NewGroupNorm(numGroups, outChannels),
		Norm2:    NewGroupNorm(numGroups, outChannels),
		TimeMLP:  NewLinear(timeEmbDim, outChannels, rng),
		SkipConv: skipConv,
		Channels: outChannels,
	}
}

// Forward computes the residual block output.
func (r *ResBlock) Forward(x, timeEmb *Tensor) *Tensor {
	batch := x.Shape[0]
	height := x.Shape[2]
	width := x.Shape[3]

	// First conv + norm + activation
	h := r.Conv1.Forward(x)
	h = r.Norm1.Forward(h)
	h = h.SiLU()

	// Add time embedding
	timeProj := r.TimeMLP.Forward(timeEmb)
	timeProj = timeProj.SiLU()

	// Broadcast time projection to spatial dimensions
	// timeProj shape: [batch, channels] -> need to add to [batch, channels, height, width]
	for b := 0; b < batch; b++ {
		for c := 0; c < r.Channels; c++ {
			timeVal := timeProj.At(b, c)
			for hi := 0; hi < height; hi++ {
				for wi := 0; wi < width; wi++ {
					val := getVal4D(h, b, c, hi, wi)
					setVal4D(h, val+timeVal, b, c, hi, wi)
				}
			}
		}
	}

	// Second conv + norm + activation
	h = r.Conv2.Forward(h)
	h = r.Norm2.Forward(h)
	h = h.SiLU()

	// Skip connection
	skip := x
	if r.SkipConv != nil {
		skip = r.SkipConv.Forward(x)
	}

	return h.Add(skip)
}

// ParameterCount returns total parameters.
func (r *ResBlock) ParameterCount() int {
	count := r.Conv1.ParameterCount() + r.Conv2.ParameterCount()
	count += r.Norm1.ParameterCount() + r.Norm2.ParameterCount()
	count += r.TimeMLP.ParameterCount()
	if r.SkipConv != nil {
		count += r.SkipConv.ParameterCount()
	}
	return count
}

// Downsample reduces spatial dimensions by 2.
type Downsample struct {
	Conv *Conv2d
}

// NewDownsample creates a downsample block.
func NewDownsample(channels int, rng *rand.Rand) *Downsample {
	return &Downsample{
		Conv: NewConv2d(channels, channels, 3, 2, 1, rng),
	}
}

// Forward downsamples the input.
func (d *Downsample) Forward(x *Tensor) *Tensor {
	return d.Conv.Forward(x)
}

// Upsample increases spatial dimensions by 2.
type Upsample struct {
	Conv *Conv2d
}

// NewUpsample creates an upsample block.
func NewUpsample(channels int, rng *rand.Rand) *Upsample {
	return &Upsample{
		Conv: NewConv2d(channels, channels, 3, 1, 1, rng),
	}
}

// Forward upsamples the input using nearest neighbor interpolation.
func (u *Upsample) Forward(x *Tensor) *Tensor {
	batch := x.Shape[0]
	channels := x.Shape[1]
	height := x.Shape[2]
	width := x.Shape[3]

	newHeight := height * 2
	newWidth := width * 2

	upsampled := Zeros([]int{batch, channels, newHeight, newWidth})

	for b := 0; b < batch; b++ {
		for c := 0; c < channels; c++ {
			for h := 0; h < height; h++ {
				for w := 0; w < width; w++ {
					val := getVal4D(x, b, c, h, w)
					// Copy to 2x2 block
					setVal4D(upsampled, val, b, c, h*2, w*2)
					setVal4D(upsampled, val, b, c, h*2+1, w*2)
					setVal4D(upsampled, val, b, c, h*2, w*2+1)
					setVal4D(upsampled, val, b, c, h*2+1, w*2+1)
				}
			}
		}
	}

	return u.Conv.Forward(upsampled)
}

// UNet is a simplified U-Net for noise prediction.
type UNet struct {
	// Initial projection
	InitConv *Conv2d

	// Time embedding
	TimeEmb1   *Linear
	TimeEmb2   *Linear
	TimeEmbDim int

	// Encoder
	DownBlocks  []*ResBlock
	Downsamples []*Downsample

	// Middle
	MidBlock1 *ResBlock
	MidBlock2 *ResBlock

	// Decoder
	UpBlocks  []*ResBlock
	Upsamples []*Upsample

	// Final projection
	FinalNorm *GroupNorm
	FinalConv *Conv2d

	rng *rand.Rand
}

// NewUNet creates a new U-Net.
func NewUNet(
	inChannels int,
	modelChannels int,
	channelMult []int,
	numResBlocks int,
	rng *rand.Rand,
) *UNet {
	timeEmbDim := modelChannels * 4

	unet := &UNet{
		TimeEmbDim: timeEmbDim,
		rng:        rng,
	}

	// Time embedding MLPs
	unet.TimeEmb1 = NewLinear(modelChannels, timeEmbDim, rng)
	unet.TimeEmb2 = NewLinear(timeEmbDim, timeEmbDim, rng)

	// Initial conv
	unet.InitConv = NewConv2d(inChannels, modelChannels, 3, 1, 1, rng)

	// Build encoder
	ch := modelChannels
	for level, mult := range channelMult {
		outCh := modelChannels * mult

		for i := 0; i < numResBlocks; i++ {
			unet.DownBlocks = append(unet.DownBlocks, NewResBlock(ch, outCh, timeEmbDim, rng))
			ch = outCh
		}

		if level < len(channelMult)-1 {
			unet.Downsamples = append(unet.Downsamples, NewDownsample(ch, rng))
		}
	}

	// Middle blocks
	unet.MidBlock1 = NewResBlock(ch, ch, timeEmbDim, rng)
	unet.MidBlock2 = NewResBlock(ch, ch, timeEmbDim, rng)

	// Build decoder (reverse of encoder)
	for level := len(channelMult) - 1; level >= 0; level-- {
		mult := channelMult[level]
		outCh := modelChannels * mult

		for i := 0; i < numResBlocks; i++ {
			// Double channels for skip connections
			skipCh := ch
			if i == 0 && level < len(channelMult)-1 {
				skipCh = ch * 2
			}
			_ = skipCh
			unet.UpBlocks = append(unet.UpBlocks, NewResBlock(ch, outCh, timeEmbDim, rng))
			ch = outCh
		}

		if level > 0 {
			unet.Upsamples = append(unet.Upsamples, NewUpsample(ch, rng))
		}
	}

	// Final layers
	numGroups := 8
	if ch < 8 {
		numGroups = ch
	}
	unet.FinalNorm = NewGroupNorm(numGroups, ch)
	unet.FinalConv = NewConv2d(ch, inChannels, 3, 1, 1, rng)

	return unet
}

// Forward performs the U-Net forward pass.
func (u *UNet) Forward(x *Tensor, timesteps []int) *Tensor {
	// Get timestep embedding
	timeEmb := GetTimestepEmbedding(timesteps, u.TimeEmbDim/4)
	timeEmb = u.TimeEmb1.Forward(timeEmb)
	timeEmb = timeEmb.SiLU()
	timeEmb = u.TimeEmb2.Forward(timeEmb)

	// Initial conv
	h := u.InitConv.Forward(x)

	// Store for skip connections
	skips := []*Tensor{h.Clone()}

	// Encoder
	blockIdx := 0
	downIdx := 0
	for level := 0; level < len(u.Downsamples)+1; level++ {
		numBlocks := 1 // Simplified
		for i := 0; i < numBlocks && blockIdx < len(u.DownBlocks); i++ {
			h = u.DownBlocks[blockIdx].Forward(h, timeEmb)
			skips = append(skips, h.Clone())
			blockIdx++
		}
		if downIdx < len(u.Downsamples) {
			h = u.Downsamples[downIdx].Forward(h)
			skips = append(skips, h.Clone())
			downIdx++
		}
	}

	// Middle
	h = u.MidBlock1.Forward(h, timeEmb)
	h = u.MidBlock2.Forward(h, timeEmb)

	// Decoder (simplified - not using full skip connections for clarity)
	upBlockIdx := 0
	upIdx := 0
	for level := 0; level < len(u.Upsamples)+1; level++ {
		numBlocks := 1
		for i := 0; i < numBlocks && upBlockIdx < len(u.UpBlocks); i++ {
			h = u.UpBlocks[upBlockIdx].Forward(h, timeEmb)
			upBlockIdx++
		}
		if upIdx < len(u.Upsamples) {
			h = u.Upsamples[upIdx].Forward(h)
			upIdx++
		}
	}

	// Final
	h = u.FinalNorm.Forward(h)
	h = h.SiLU()
	h = u.FinalConv.Forward(h)

	return h
}

// ParameterCount returns total parameters in the U-Net.
func (u *UNet) ParameterCount() int {
	count := u.InitConv.ParameterCount()
	count += u.TimeEmb1.ParameterCount() + u.TimeEmb2.ParameterCount()

	for _, b := range u.DownBlocks {
		count += b.ParameterCount()
	}
	for _, d := range u.Downsamples {
		count += d.Conv.ParameterCount()
	}

	count += u.MidBlock1.ParameterCount() + u.MidBlock2.ParameterCount()

	for _, b := range u.UpBlocks {
		count += b.ParameterCount()
	}
	for _, up := range u.Upsamples {
		count += up.Conv.ParameterCount()
	}

	count += u.FinalNorm.ParameterCount() + u.FinalConv.ParameterCount()

	return count
}

File: mini-diffusion/src/unet.rs
//! U-Net Architecture - The noise prediction network
//!
//! A simplified U-Net for image denoising, similar to what Stable Diffusion uses
//! but much smaller for educational purposes.

use crate::nn::{Conv2d, GroupNorm, Linear};
use crate::tensor::Tensor;
use crate::diffusion::get_timestep_embedding;

/// Residual Block with GroupNorm and time embedding
#[derive(Debug, Clone)]
pub struct ResBlock {
    pub conv1: Conv2d,
    pub conv2: Conv2d,
    pub norm1: GroupNorm,
    pub norm2: GroupNorm,
    pub time_mlp: Linear,
    pub skip_conv: Option<Conv2d>,
    pub channels: usize,
}

impl ResBlock {
    pub fn new(in_channels: usize, out_channels: usize, time_emb_dim: usize) -> Self {
        let skip_conv = if in_channels != out_channels {
            Some(Conv2d::new(in_channels, out_channels, 1, 1, 0))
        } else {
            None
        };

        ResBlock {
            conv1: Conv2d::new(in_channels, out_channels, 3, 1, 1),
            conv2: Conv2d::new(out_channels, out_channels, 3, 1, 1),
            norm1: GroupNorm::new(8.min(out_channels), out_channels),
            norm2: GroupNorm::new(8.min(out_channels), out_channels),
            time_mlp: Linear::new(time_emb_dim, out_channels),
            skip_conv,
            channels: out_channels,
        }
    }

    pub fn forward(&self, x: &Tensor, time_emb: &Tensor) -> Tensor {
        let shape = x.shape();
        let batch = shape[0];
        let height = shape[2];
        let width = shape[3];

        // First conv + norm
        let h = self.conv1.forward(x);
        let h = self.norm1.forward(&h);
        let h = h.silu();

        // Add time embedding
        // time_emb shape: [batch, time_emb_dim]
        // Need to project and add to spatial features
        let time_proj = self.time_mlp.forward(time_emb);
        let time_proj = time_proj.silu();
        
        // Broadcast time projection to spatial dimensions
        // shape: [batch, channels] -> [batch, channels, height, width]
        let time_proj = time_proj.reshape(&[batch, self.channels, 1, 1]);
        let time_proj = time_proj.broadcast(&[batch, self.channels, height, width]);
        
        let h = h.add(&time_proj);

        // Second conv + norm
        let h = self.conv2.forward(&h);
        let h = self.norm2.forward(&h);
        let h = h.silu();

        // Skip connection
        let skip = match &self.skip_conv {
            Some(conv) => conv.forward(x),
            None => x.clone(),
        };

        h.add(&skip)
    }

    pub fn parameters(&self) -> Vec<&Tensor> {
        let mut params = self.conv1.parameters();
        params.extend(self.conv2.parameters());
        params.extend(self.norm1.parameters());
        params.extend(self.norm2.parameters());
        params.extend(self.time_mlp.parameters());
        if let Some(skip) = &self.skip_conv {
            params.extend(skip.parameters());
        }
        params
    }
}

/// Downsampling block (stride 2 conv)
#[derive(Debug, Clone)]
pub struct Downsample {
    pub conv: Conv2d,
}

impl Downsample {
    pub fn new(channels: usize) -> Self {
        Downsample {
            conv: Conv2d::new(channels, channels, 3, 2, 1),
        }
    }

    pub fn forward(&self, x: &Tensor) -> Tensor {
        self.conv.forward(x)
    }

    pub fn parameters(&self) -> Vec<&Tensor> {
        self.conv.parameters()
    }
}

/// Upsampling block (nearest neighbor + conv)
#[derive(Debug, Clone)]
pub struct Upsample {
    pub conv: Conv2d,
}

impl Upsample {
    pub fn new(channels: usize) -> Self {
        Upsample {
            conv: Conv2d::new(channels, channels, 3, 1, 1),
        }
    }

    pub fn forward(&self, x: &Tensor) -> Tensor {
        let shape = x.shape();
        let batch = shape[0];
        let channels = shape[1];
        let height = shape[2];
        let width = shape[3];

        // Nearest neighbor upsampling 2x
        let new_height = height * 2;
        let new_width = width * 2;
        
        let mut upsampled = Tensor::zeros(&[batch, channels, new_height, new_width]);
        let x_contig = x.to_contiguous();
        let x_data = x_contig.as_slice();
        let up_data = upsampled.as_mut_slice();

        for b in 0..batch {
            for c in 0..channels {
                for h in 0..height {
                    for w in 0..width {
                        let val = x_data[b * channels * height * width + c * height * width + h * width + w];
                        // Copy to 2x2 block
                        for dh in 0..2 {
                            for dw in 0..2 {
                                let nh = h * 2 + dh;
                                let nw = w * 2 + dw;
                                up_data[b * channels * new_height * new_width + c * new_height * new_width + nh * new_width + nw] = val;
                            }
                        }
                    }
                }
            }
        }

        self.conv.forward(&upsampled)
    }

    pub fn parameters(&self) -> Vec<&Tensor> {
        self.conv.parameters()
    }
}

/// Simplified U-Net for diffusion models
/// 
/// Architecture:
/// - Encoder: progressively downsamples while increasing channels
/// - Middle: bottleneck with attention
/// - Decoder: progressively upsamples while decreasing channels
/// - Skip connections between encoder and decoder
#[derive(Debug)]
pub struct UNet {
    // Time embedding
    pub time_mlp1: Linear,
    pub time_mlp2: Linear,
    pub time_emb_dim: usize,
    
    // Initial convolution
    pub init_conv: Conv2d,
    
    // Encoder blocks
    pub down1: ResBlock,
    pub down2: ResBlock,
    pub down3: ResBlock,
    pub pool1: Downsample,
    pub pool2: Downsample,
    
    // Middle (bottleneck)
    pub mid1: ResBlock,
    pub mid2: ResBlock,
    
    // Decoder blocks
    pub up3: ResBlock,
    pub up2: ResBlock,
    pub up1: ResBlock,
    pub upsample2: Upsample,
    pub upsample1: Upsample,
    
    // Final convolution
    pub final_norm: GroupNorm,
    pub final_conv: Conv2d,
    
    // Channel configuration
    pub channels: [usize; 4],
}

impl UNet {
    /// Create a new U-Net
    /// 
    /// Args:
    ///   in_channels: Input image channels (e.g., 3 for RGB)
    ///   model_channels: Base number of channels (multiplied at each level)
    ///   out_channels: Output channels (same as input for noise prediction)
    pub fn new(in_channels: usize, model_channels: usize, out_channels: usize) -> Self {
        let time_emb_dim = model_channels * 4;
        
        // Channel progression: model_channels -> 2x -> 4x -> 4x
        let channels = [
            model_channels,
            model_channels * 2,
            model_channels * 4,
            model_channels * 4,
        ];

        UNet {
            // Time embedding MLP
            time_mlp1: Linear::new(model_channels, time_emb_dim),
            time_mlp2: Linear::new(time_emb_dim, time_emb_dim),
            time_emb_dim,
            
            // Initial conv: in_channels -> model_channels
            init_conv: Conv2d::new(in_channels, channels[0], 3, 1, 1),
            
            // Encoder
            down1: ResBlock::new(channels[0], channels[1], time_emb_dim),
            pool1: Downsample::new(channels[1]),
            down2: ResBlock::new(channels[1], channels[2], time_emb_dim),
            pool2: Downsample::new(channels[2]),
            down3: ResBlock::new(channels[2], channels[3], time_emb_dim),
            
            // Middle
            mid1: ResBlock::new(channels[3], channels[3], time_emb_dim),
            mid2: ResBlock::new(channels[3], channels[3], time_emb_dim),
            
            // Decoder (note: channels doubled for skip connections)
            up3: ResBlock::new(channels[3] + channels[3], channels[2], time_emb_dim),
            upsample2: Upsample::new(channels[2]),
            up2: ResBlock::new(channels[2] + channels[2], channels[1], time_emb_dim),
            upsample1: Upsample::new(channels[1]),
            up1: ResBlock::new(channels[1] + channels[1], channels[0], time_emb_dim),
            
            // Final
            final_norm: GroupNorm::new(8.min(channels[0]), channels[0]),
            final_conv: Conv2d::new(channels[0], out_channels, 3, 1, 1),
            
            channels,
        }
    }

    /// Forward pass
    /// 
    /// Args:
    ///   x: Noisy image [batch, channels, height, width]
    ///   timesteps: Current timestep for each batch element
    /// 
    /// Returns:
    ///   Predicted noise [batch, channels, height, width]
    pub fn forward(&self, x: &Tensor, timesteps: &[usize]) -> Tensor {
        let _batch = x.shape()[0];
        
        // 1. Time embedding
        let t_emb = get_timestep_embedding(timesteps, self.channels[0]);
        let t_emb = self.time_mlp1.forward(&t_emb);
        let t_emb = t_emb.silu();
        let t_emb = self.time_mlp2.forward(&t_emb);
        
        // 2. Initial convolution
        let h = self.init_conv.forward(x);
        
        // 3. Encoder (save for skip connections)
        let h1 = self.down1.forward(&h, &t_emb);      // [batch, c1, H, W]
        let h = self.pool1.forward(&h1);               // [batch, c1, H/2, W/2]
        
        let h2 = self.down2.forward(&h, &t_emb);      // [batch, c2, H/2, W/2]
        let h = self.pool2.forward(&h2);               // [batch, c2, H/4, W/4]
        
        let h3 = self.down3.forward(&h, &t_emb);      // [batch, c3, H/4, W/4]
        
        // 4. Middle
        let h = self.mid1.forward(&h3, &t_emb);
        let h = self.mid2.forward(&h, &t_emb);
        
        // 5. Decoder with skip connections
        let h = Tensor::concat(&[&h, &h3], 1);         // Concatenate along channel dim
        let h = self.up3.forward(&h, &t_emb);
        let h = self.upsample2.forward(&h);
        
        let h = Tensor::concat(&[&h, &h2], 1);
        let h = self.up2.forward(&h, &t_emb);
        let h = self.upsample1.forward(&h);
        
        let h = Tensor::concat(&[&h, &h1], 1);
        let h = self.up1.forward(&h, &t_emb);
        
        // 6. Final output
        let h = self.final_norm.forward(&h);
        let h = h.silu();
        self.final_conv.forward(&h)
    }

    /// Get all trainable parameters
    pub fn parameters(&self) -> Vec<&Tensor> {
        let mut params = vec![];
        
        params.extend(self.time_mlp1.parameters());
        params.extend(self.time_mlp2.parameters());
        params.extend(self.init_conv.parameters());
        
        params.extend(self.down1.parameters());
        params.extend(self.down2.parameters());
        params.extend(self.down3.parameters());
        params.extend(self.pool1.parameters());
        params.extend(self.pool2.parameters());
        
        params.extend(self.mid1.parameters());
        params.extend(self.mid2.parameters());
        
        params.extend(self.up3.parameters());
        params.extend(self.up2.parameters());
        params.extend(self.up1.parameters());
        params.extend(self.upsample2.parameters());
        params.extend(self.upsample1.parameters());
        
        params.extend(self.final_norm.parameters());
        params.extend(self.final_conv.parameters());
        
        params
    }

    /// Count total number of parameters
    pub fn num_parameters(&self) -> usize {
        self.parameters().iter().map(|p| p.numel()).sum()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_resblock() {
        let block = ResBlock::new(32, 64, 128);
        let x = Tensor::randn(&[2, 32, 16, 16]);
        let t = Tensor::randn(&[2, 128]);
        let y = block.forward(&x, &t);
        assert_eq!(y.shape(), &[2, 64, 16, 16]);
    }

    #[test]
    fn test_downsample() {
        let down = Downsample::new(64);
        let x = Tensor::randn(&[2, 64, 32, 32]);
        let y = down.forward(&x);
        assert_eq!(y.shape(), &[2, 64, 16, 16]);
    }

    #[test]
    fn test_upsample() {
        let up = Upsample::new(64);
        let x = Tensor::randn(&[2, 64, 16, 16]);
        let y = up.forward(&x);
        assert_eq!(y.shape(), &[2, 64, 32, 32]);
    }

    #[test]
    fn test_unet_forward() {
        let unet = UNet::new(3, 32, 3);
        let x = Tensor::randn(&[1, 3, 32, 32]);
        let timesteps = vec![100];
        let y = unet.forward(&x, &timesteps);
        assert_eq!(y.shape(), &[1, 3, 32, 32]);
        
        println!("UNet parameters: {}", unet.num_parameters());
    }
}
